File: android/app/src/main/java/app/remotex/ui/screens/session/events/DiffView.kt
package app.remotex.ui.screens.session.events

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.remotex.ui.theme.Ink
import app.remotex.ui.theme.InkDim
import app.remotex.ui.theme.Line
import app.remotex.ui.theme.LocalPalette
import app.remotex.ui.theme.Ok
import app.remotex.ui.theme.Warn

/**
 * Edit diffs the way Claude Code / Codex show them: one card per file with
 * a verb + path header, +N/−M counts, and green/red-tinted lines.
 *
 * Input contract (services/daemon/adapters/items.py::_format_changes):
 *   summary — one "verb /path" line per file (arrives as Tool.command)
 *   diff    — single file: the raw unified diff;
 *             multiple files: sections prefixed with "--- /path".
 */
private data class DiffFile(
    val verb: String,
    val path: String,
    val movedTo: String?,
    val lines: List<String>,
)

private const val COLLAPSED_LINES = 14

private val VERB_RE = Regex("""^(\w+)\s+(.+?)(?:\s+→\s+(.+))?$""")

private fun parseFiles(summary: String, diff: String): List<DiffFile> {
    val verbs = summary.split('\n').mapNotNull { raw ->
        VERB_RE.find(raw.trim())?.let { m ->
            Triple(m.groupValues[1], m.groupValues[2], m.groupValues.getOrNull(3)?.ifBlank { null })
        }
    }
    val allLines = diff.split('\n')
    if (verbs.size <= 1) {
        val v = verbs.firstOrNull()
        return listOf(
            DiffFile(v?.first ?: "update", v?.second ?: "", v?.third, allLines),
        )
    }
    val files = mutableListOf<DiffFile>()
    var current: MutableList<String>? = null
    for (line in allLines) {
        val head = if (line.startsWith("--- ")) line.removePrefix("--- ") else null
        val match = head?.let { h -> verbs.find { it.second == h } }
        if (match != null) {
            val lines = mutableListOf<String>()
            files.add(DiffFile(match.first, match.second, match.third, lines))
            current = lines
            continue
        }
        current?.add(line)
    }
    return files.ifEmpty { listOf(DiffFile("update", "", null, allLines)) }
}

@Composable
internal fun DiffView(summary: String, diff: String, streaming: Boolean) {
    val files = parseFiles(summary, diff)
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        files.forEachIndexed { index, file ->
            FileDiff(file, streaming, key = "$index-${file.path}")
        }
    }
}

@Composable
private fun FileDiff(file: DiffFile, streaming: Boolean, key: String) {
    var expanded by rememberSaveable(key) { mutableStateOf(false) }
    val adds = file.lines.count { it.startsWith("+") && !it.startsWith("+++") }
    val dels = file.lines.count { it.startsWith("-") && !it.startsWith("---") }
    val overflow = file.lines.size > COLLAPSED_LINES
    val shown = when {
        expanded || !overflow -> file.lines
        // While streaming, the tail is the part being written.
        streaming -> file.lines.takeLast(COLLAPSED_LINES)
        else -> file.lines.take(COLLAPSED_LINES)
    }
    val palette = LocalPalette.current

    Column(
        Modifier
            .fillMaxWidth()
            .background(palette.panel2.copy(alpha = 0.55f), RoundedCornerShape(6.dp)),
    ) {
        Row(
            Modifier
                .fillMaxWidth()
                .background(palette.panel2, RoundedCornerShape(6.dp, 6.dp, 0.dp, 0.dp))
                .clickable(enabled = overflow) { expanded = !expanded }
                .padding(horizontal = 9.dp, vertical = 5.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Text(file.verb, color = InkDim, fontFamily = FontFamily.Monospace, fontSize = 11.sp)
            Text(
                file.path + (file.movedTo?.let { " → $it" } ?: ""),
                color = Ink,
                fontFamily = FontFamily.Monospace,
                fontSize = 11.sp,
                maxLines = 1,
                // Head ellipsis keeps the filename (the useful end) visible.
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.weight(1f, fill = false),
            )
            Spacer(Modifier.weight(1f))
            if (adds > 0) {
                Text("+$adds", color = Ok, fontFamily = FontFamily.Monospace, fontSize = 11.sp)
            }
            if (dels > 0) {
                Text("−$dels", color = Warn, fontFamily = FontFamily.Monospace, fontSize = 11.sp)
            }
            if (overflow) {
                Text(
                    if (expanded) "collapse" else "expand",
                    color = InkDim,
                    fontFamily = FontFamily.Monospace,
                    fontSize = 10.sp,
                )
            }
        }
        Column(
            Modifier
                .fillMaxWidth()
                .horizontalScroll(rememberScrollState())
                .padding(vertical = 4.dp),
        ) {
            shown.forEach { line ->
                val bg = when {
                    line.startsWith("+") && !line.startsWith("+++") -> palette.ok.copy(alpha = 0.10f)
                    line.startsWith("-") && !line.startsWith("---") -> palette.warn.copy(alpha = 0.09f)
                    else -> androidx.compose.ui.graphics.Color.Transparent
                }
                val fg = when {
                    line.startsWith("+++") || line.startsWith("---") -> palette.inkDim
                    line.startsWith("@@") -> palette.accentDeep
                    line.startsWith("+") -> palette.ok
                    line.startsWith("-") -> palette.warn
                    else -> palette.inkDim
                }
                Text(
                    line.ifEmpty { " " },
                    color = fg,
                    fontFamily = FontFamily.Monospace,
                    fontSize = 11.sp,
                    maxLines = 1,
                    softWrap = false,
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(bg)
                        .padding(horizontal = 9.dp),
                )
            }
            if (!expanded && overflow) {
                Text(
                    "… ${file.lines.size - COLLAPSED_LINES} more lines",
                    color = palette.accent,
                    fontFamily = FontFamily.Monospace,
                    fontSize = 10.sp,
                    fontWeight = FontWeight.Normal,
                    modifier = Modifier
                        .clickable { expanded = true }
                        .padding(horizontal = 9.dp, vertical = 2.dp),
                )
            }
        }
    }
}

File: apple/Remotex/DiffView.swift
import SwiftUI

// Edit diffs the way Claude Code / Codex show them: one card per file with
// verb + path header, +N/−M counts, and green/red-tinted lines.
//
// Input contract (services/daemon/adapters/items.py::_format_changes):
//   summary — one "verb /path" line per file (arrives as the tool's detail)
//   diff    — single file: the raw unified diff;
//             multiple files: sections prefixed with "--- /path".

private struct DiffFile: Identifiable {
    let id = UUID()
    let verb: String
    let path: String
    let movedTo: String?
    let lines: [String]
}

private let collapsedLines = 14

private func parseDiffFiles(summary: String, diff: String) -> [DiffFile] {
    let verbs: [(verb: String, path: String, movedTo: String?)] = summary
        .components(separatedBy: "\n")
        .compactMap { raw in
            let line = raw.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { return nil }
            let parts = line.split(separator: " ", maxSplits: 1).map(String.init)
            guard parts.count == 2 else { return nil }
            let rest = parts[1]
            if let arrow = rest.range(of: " → ") {
                return (parts[0], String(rest[..<arrow.lowerBound]), String(rest[arrow.upperBound...]))
            }
            return (parts[0], rest, nil)
        }

    let allLines = diff.components(separatedBy: "\n")
    guard verbs.count > 1 else {
        let v = verbs.first
        return [DiffFile(
            verb: v?.verb ?? "update",
            path: v?.path ?? "",
            movedTo: v?.movedTo,
            lines: allLines
        )]
    }

    var files: [DiffFile] = []
    var currentVerb: (verb: String, path: String, movedTo: String?)?
    var buffer: [String] = []
    func flush() {
        guard let v = currentVerb else { return }
        files.append(DiffFile(verb: v.verb, path: v.path, movedTo: v.movedTo, lines: buffer))
        buffer = []
    }
    for line in allLines {
        if line.hasPrefix("--- ") {
            let path = String(line.dropFirst(4))
            if let match = verbs.first(where: { $0.path == path }) {
                flush()
                currentVerb = match
                continue
            }
        }
        buffer.append(line)
    }
    flush()
    return files.isEmpty
        ? [DiffFile(verb: "update", path: "", movedTo: nil, lines: allLines)]
        : files
}

struct DiffView: View {
    let summary: String
    let diff: String
    let streaming: Bool

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            ForEach(parseDiffFiles(summary: summary, diff: diff)) { file in
                FileDiffCard(file: file, streaming: streaming)
            }
        }
    }
}

private struct FileDiffCard: View {
    fileprivate let file: DiffFile
    let streaming: Bool
    @State private var expanded = false

    private var adds: Int {
        file.lines.filter { $0.hasPrefix("+") && !$0.hasPrefix("+++") }.count
    }
    private var dels: Int {
        file.lines.filter { $0.hasPrefix("-") && !$0.hasPrefix("---") }.count
    }
    private var overflow: Bool { file.lines.count > collapsedLines }
    private var shown: [String] {
        if expanded || !overflow { return file.lines }
        // While streaming, the tail is the part being written.
        return streaming ? Array(file.lines.suffix(collapsedLines)) : Array(file.lines.prefix(collapsedLines))
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Button {
                if overflow { expanded.toggle() }
            } label: {
                HStack(spacing: 8) {
                    Text(file.verb)
                        .foregroundStyle(Color.remotexMuted)
                    Text(file.path + (file.movedTo.map { " → \($0)" } ?? ""))
                        .foregroundStyle(Color.remotexText)
                        .lineLimit(1)
                        .truncationMode(.head)
                    Spacer(minLength: 4)
                    if adds > 0 {
                        Text("+\(adds)").foregroundStyle(Color.remotexGreen)
                    }
                    if dels > 0 {
                        Text("−\(dels)").foregroundStyle(Color.remotexWarn)
                    }
                    if overflow {
                        Text(expanded ? "collapse" : "expand")
                            .foregroundStyle(Color.remotexMuted)
                    }
                }
                .font(.system(size: 11, design: .monospaced))
                .padding(.horizontal, 9)
                .padding(.vertical, 5)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(Color.remotexSurface2)
            }
            .buttonStyle(.plain)

            ScrollView(.horizontal, showsIndicators: false) {
                VStack(alignment: .leading, spacing: 0) {
                    ForEach(Array(shown.enumerated()), id: \.offset) { _, line in
                        Text(line.isEmpty ? " " : line)
                            .font(.system(size: 11, design: .monospaced))
                            .foregroundStyle(color(for: line))
                            .padding(.horizontal, 9)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(background(for: line))
                    }
                    if !expanded, overflow {
                        Button("… \(file.lines.count - collapsedLines) more lines") {
                            expanded = true
                        }
                        .font(.system(size: 10, design: .monospaced))
                        .foregroundStyle(Color.remotexAccent)
                        .padding(.horizontal, 9)
                        .padding(.vertical, 2)
                    }
                }
                .padding(.vertical, 4)
            }
        }
        .background(Color.remotexSurface2.opacity(0.55))
        .clipShape(RoundedRectangle(cornerRadius: 6))
    }

    private func color(for line: String) -> Color {
        if line.hasPrefix("+++") || line.hasPrefix("---") { return .remotexMuted }
        if line.hasPrefix("@@") { return .remotexAccentDeep }
        if line.hasPrefix("+") { return .remotexGreen }
        if line.hasPrefix("-") { return .remotexWarn }
        return .remotexMuted
    }

    private func background(for line: String) -> Color {
        if line.hasPrefix("+"), !line.hasPrefix("+++") { return Color.remotexGreen.opacity(0.10) }
        if line.hasPrefix("-"), !line.hasPrefix("---") { return Color.remotexWarn.opacity(0.09) }
        return .clear
    }
}
